### PR TITLE
Fix line numbers for thrift tests

### LINKED ISSUE #229: `thrift_test` loses test line numbers
I don't know if it's easy to fix this and it shouldn't hold us back I don't think, but I thought I'd make an issue for discussion's sake (maybe someone has an idea for a quick fix).  The `thrift_test` macro makes it impossible to run, e.g., `mix test test/thrift/generator/binary_protocol_test.exs:599` because the line numbers get lost.  Maybe it's as simple as adding `location: :keep` to the `defmacro`?

### PR DESCRIPTION
**Pull request by fishcakez**
Create/register tests inline in test module so that test line numbers can be used with `mix test`.

Closes #229.

### CODE REVIEW

File: test/support/lib/thrift_test_case.ex [hunk 1]
```diff
@@ -1,131 +1,134 @@
...
+    filename
+  end
+
+  defp require_file(file, dir) do
+    case Code.require_file(file, dir) do
+      nil ->
+        []
+      modules ->
+        modules = Enum.map(modules, fn({module, _}) -> Module.split(module) end)
+        Enum.filter_map(modules, &alias?(&1, modules), &Module.concat/1)
```
Comment (jparise) [inline test/support/lib/thrift_test_case.ex]: Or (shorter and doesn't rebind `modules`):
```elixir
modules
|> Enum.map(fn({module, _}) -> Module.split(module) end)
|> Enum.filter_map(&alias?(&1, modules), &Module.concat/1)
```

File: test/support/lib/thrift_test_case.ex [hunk 2]
```diff
@@ -142,22 +145,52 @@ defmodule ThriftTestCase do
...
+    module = erlang_module(source_file)
+    if loaded?(source_file, module) do
+      module
+    else
+      erlang_compile(source_file)
+    end
+  end
+
+  defp loaded?(source_file \\ nil, mod) do
+    source_file = source_file && String.to_char_list(source_file)
```
Comment (jparise) [inline test/support/lib/thrift_test_case.ex]: Isn't the `String.to_charlist/1` spelling preferred for Elixir 1.3+?

**Comment by jparise** [comment]
One related comment is that the Erlang bits are currently useful (for integration testing), but I hope that we won't need them after too long once we're standing firmly on our own legs.

For binary compatibility testing, I think we should generate some "golden" test data files and write them as both serialized Thrift data and their equivalent Erlang term format. That will allow us to validate our serialization/deserialization code without using any of the Erlang runtime.

File: test/support/lib/thrift_test_case.ex [hunk 3]
```diff
@@ -1,131 +1,134 @@
...
+      nil ->
+        []
+      modules ->
+        modules = Enum.map(modules, fn({module, _}) -> Module.split(module) end)
+        Enum.filter_map(modules, &alias?(&1, modules), &Module.concat/1)
+      end
+  end
+
+  defp alias?(module, modules) do
+    # alias when module when parent in namespace does not exist
```
Comment (fishcakez) [inline test/support/lib/thrift_test_case.ex]: That's supposed to be `when parent module`

File: test/support/lib/thrift_test_case.ex (hunk 1 above)
Comment (fishcakez) [inline test/support/lib/thrift_test_case.ex]: That isn't quite equivalent, but it just shows how confusing that piece of code is. We need the rebound `modules` as second argument to `alias?/2`.

File: test/support/lib/thrift_test_case.ex (hunk 2 above)
Comment (fishcakez) [inline test/support/lib/thrift_test_case.ex]: Yes it is, didn't update from the prior code.

**Pull request merged**